src/ecdsa/Der.mo

```diff
@@ -11,6 +14,101 @@ module {
   type DerSignature = Types.DerSignature;
   type Signature = Types.Signature;
 
+  // Serialize signature to DER format:
+  // 0x30 [total-length] 0x02 [R-length] [R] 0x02 [S-length] [S]
+  func _encodeSignature(r : [Nat8], s : [Nat8]) : DerSignature {
+    // Specified value is serialized in big endian with all prefix zeroes
+    // stripped, except if the most significant bit of the first non-zero byte
+    // is set, then we prepend one zero so that the value is not interpreted as
+    // negative.
+    func preprocessData(data : [Nat8]) : [Nat8] {
+      // Serialize value big endian to 32 byte array.
+      let outputBuf = Buffer.Buffer<Nat8>(data.size());
+
+      for (i in Iter.range(0, data.size() - 1)) {
+        // We are looking for the first non-zero byte.
+        if (outputBuf.size() == 0) {
+          // Check whether the current byte is a non-zero.
+          if (data[i] != 0) {
+            // Check whether the current byte has its msb set.
+            if (data[i] >= 0x80) {
+              // Msb is set, add zero to output buffer.
+              outputBuf.add(0x00);
+            };
+            // Add the current byte.
+            outputBuf.add(data[i]);
+          };
+        } else {
+          // We already found the first non-zero byte and we're just copying
+          // subequent bytes into the output buffer.
+          outputBuf.add(data[i]);
+        };
+      };
+
+      return outputBuf.toArray();
+    };
+
+    let output = Buffer.Buffer<Nat8>(0);
+    let rData : [Nat8] = preprocessData(r);
+    let sData : [Nat8] = preprocessData(s);
+
+    // Add DER identifier.
+    output.add(0x30);
+    // Total size of everything that comes next, excluding sighash type.
+    output.add(Nat8.fromIntWrap(
+      // DER Sequence identifier: 0x02.
+      1
+      // Signature r component size.
+      + 1
+      // Signature r component.
+      + rData.size()
+      // DER Sequence identifier : 0x02.
+      + 1
+      // Signature s component size.
+      + 1
+      // Signature s component.
+      + sData.size()
+    ));
+    // DER sequence identifier.
+    output.add(0x02);
+    // Signature r component size.
+    output.add(Nat8.fromIntWrap(rData.size()));
+
+    // Signature r component.
+    for (i in rData.vals()) {
+      output.add(i);
+    };
+
+    // DER sequence identifier.
+    output.add(0x02);
+    // Signature s component size.
+    output.add(Nat8.fromIntWrap(sData.size()));
+
+    // Signature s component.
+    for (i in sData.vals()) {
+      output.add(i);
+    };
+
+    // sighashtype
+    // output.add(Nat8.fromNat(Nat32.toNat(sighashType)));
+
```

Comment:
I would be in favor or removing these commented instructions, as we discussed they don't have to be here. Furthermore git allows us to keep history while not having to keep commented all our changes.

Reply:
Wasn't intentional, will remove.